src/main/java/uk/gov/justice/digital/converter/Converter.java

```diff
@@ -6,32 +6,67 @@
 import org.apache.spark.sql.SparkSession;
 import org.apache.spark.sql.types.StructType;
 
+import java.util.Arrays;
+import java.util.Optional;
+
 import static org.apache.spark.sql.types.DataTypes.StringType;
 import static uk.gov.justice.digital.converter.Converter.ParsedDataFields.*;
 
 public abstract class Converter {
 
     private static final boolean NOT_NULL = false;
+    private static final boolean IS_NULL = true;
 
     // TODO - at the moment we haven't really formalised the idea of our internal format which we could do instead of
     //        defining the fields and the struct here.
 
     // Constants defining the fields used in the common output format.
     public static class ParsedDataFields {
-        public static final String DATA = "data";
-        public static final String METADATA = "metadata";
-        public static final String OPERATION = "operation";
-        public static final String SOURCE = "source";
-        public static final String TABLE = "table";
+        public static final String RAW = "raw"; // the raw incoming value, unchanged
+        public static final String DATA = "data"; // the payload. Not always present
+        public static final String METADATA = "metadata"; // the payload metadata. Always present
+
+        // From the metadata. By default we use _ as a differentiator so that payloads with
+        // metadata fields do not get overwritten
+        public static final String TIMESTAMP = "_timestamp";
+        public static final String KEY = "_key";
+        public static final String DATA_TYPE = "_datatype";
+        public static final String OPERATION = "_operation";
+        public static final String SOURCE = "_source";
+        public static final String TABLE = "_table";
+        public static final String CONVERTER = "_converter";
     }
 
+    public enum Operation {
```

Comment:
Thanks for getting this in. This one has been bugging me for a while. I'll close the ticket for this one.